src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1046 @@
+
+  cvals = step_mem->cvals;
+  Xvecs = step_mem->Xvecs;
+
+  /* Compute spectral radius and update stats */
+  if ((step_mem->newspr))
+  {
+    retval = lsrkStep_ComputeNewSprRad(ark_mem, step_mem);
+    step_mem->nsprupdates++;
+  }
+
+  /* determine the number of required stages */
+  for (int ss = 1; ss < step_mem->stagemaxlimit; ss++)
+  {
+    if ((SUNSQR(ss) + ss - 2) >= 2 * (ark_mem->h * step_mem->sprad))
+    {
+      step_mem->reqstages = SUNMAX(ss, 2);
+      break;
+    }
+  }
+  step_mem->stagemax = SUNMAX(step_mem->reqstages, step_mem->stagemax);
+
+  /* Call the full RHS if needed. If this is the first step then we may need to
+     evaluate or copy the RHS values from an  earlier evaluation (e.g., to
+     compute h0). For subsequent steps treat this RHS evaluation as an
+     evaluation at the end of the just completed step to potentially reuse
+     (FSAL methods) RHS evaluations from the end of the last step. */
+
+  if (!(ark_mem->fn_is_current))
+  {
+    retval = step_mem->fe(ark_mem->tn, ark_mem->yn, ark_mem->fn,
+                          ark_mem->user_data);
+    step_mem->nfe++;
+    ark_mem->fn_is_current = SUNTRUE;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+  }
+
+  /* A tentative solution at t+h is returned in
+     y and its slope is evaluated in temp1.  */
+
+  w1 = FOUR / ((step_mem->reqstages + TWO) * (step_mem->reqstages - ONE));
+
+  bjm2 = ONE / THREE;
+  bjm1 = bjm2;
+
+  /* Evaluate the first stage */
+  N_VScale(ONE, ark_mem->yn, ark_mem->tempv1);
+
+  mus  = w1 * bjm1;
+  cjm1 = mus;
+
+  N_VLinearSum(ONE, ark_mem->yn, ark_mem->h * mus, ark_mem->fn, ark_mem->tempv2);
+
+  /* Evaluate stages j = 2,...,step_mem->reqstages */
+  for (int j = 2; j <= step_mem->reqstages; j++)
+  {
+    temj = (j + TWO) * (j - ONE);
+    bj   = temj / (TWO * j * (j + ONE));
+    ajm1 = ONE - bjm1;
+    mu   = (TWO * j - ONE) / j * (bj / bjm1);
+    nu   = -(j - ONE) / j * (bj / bjm2);
+    mus  = w1 * mu;
+    cj   = temj * w1 / FOUR;
+
+    /* Use the ycur array for temporary storage here */
+    retval = step_mem->fe(ark_mem->tcur + ark_mem->h * cjm1, ark_mem->tempv2,
+                          ark_mem->ycur, ark_mem->user_data);
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    cvals[0] = mus * ark_mem->h;
+    Xvecs[0] = ark_mem->ycur;
+    cvals[1] = nu;
+    Xvecs[1] = ark_mem->tempv1;
+    cvals[2] = ONE - mu - nu;
+    Xvecs[2] = ark_mem->yn;
+    cvals[3] = mu;
+    Xvecs[3] = ark_mem->tempv2;
+    cvals[4] = -mus * ajm1 * ark_mem->h;
+    Xvecs[4] = ark_mem->fn;
+
+    retval = N_VLinearCombination(5, cvals, Xvecs, ark_mem->ycur);
+    if (retval != 0) { return (ARK_VECTOROP_ERR); }
+
+    // thj = mu*thjm1 + nu*thjm2 + mus*(ONE - ajm1);
+
+    /* Shift the data for the next stage */
+    if (j < step_mem->reqstages)
+    {
+      N_VScale(ONE, ark_mem->tempv2, ark_mem->tempv1);
+      N_VScale(ONE, ark_mem->ycur, ark_mem->tempv2);
```

Comment:
Again, could we remove one call to `N_VScale` per stage with a pointer swap?

Reply:
In RKC you were able to use a pointer swap instead of the second `N_VScale` -- please do the same here.